File: Tag.cpp
//Dongyao Zhu

#include "Tag.hpp"

//need to initialise static field somewhere
map<string, int> Tag :: count = map<string, int>();

/*
 * name: constructor
 * description: creates an actual Tag
 * input: Tag's information
 * output: Tag *
 */
Tag :: Tag(double tly, double tlx, double bry, double brx,
    string name, string url = "", string style = "") 
    : name(name), cls(name), url(url), style(style){
        this -> style = "\tposition: relative;\n" + this -> style;
        if(name == "p"){
            this -> style += "\tmargin: 0%;\n";
        }
        id = name + to_string(++count[name]);
        tl = make_pair<double, double>((double)tly, (double)tlx);
        br = make_pair<double, double>((double)bry, (double)brx);
        wPct = W();
        hPct = H();
}

/*
 * name: constructor
 * description: creates a wrapper Tag
 * input: information of Tags to be wrapped; if only one Tag,
        it is for expanding row.
 * output: Tag *
 */
Tag :: Tag(double tly, double tlx, double bry, double brx,
    Tag * a, Tag * b) : name("div"){
        tl = make_pair<double, double>((double)tly, (double)tlx);
        br = make_pair<double, double>((double)bry, (double)brx);
        id = "wrap" + to_string(++count["wrap"]);
        cls = "wrap";
        children.push_back(a);
        a -> parent = this;
        a -> wPct = 100 * a -> W() / W();
        a -> hPct = 100 * a -> H() / H();
        wPct = W();
        hPct = H();
        if(b){
            children.push_back(b);
            b -> parent = this;
            b -> wPct = 100 * b -> W() / W();
            b -> hPct = 100 * b -> H() / H();
        }
        style = "\tposition: relative;\n";
}

/*
 * name: destructor
 * description: dfs, also decrease name count
 * input/output: none
 */
Tag :: ~Tag(){
    count[cls]--;
    for(Tag * t : children){
        delete t;
    }
}

/*
 * name: operator <
 * description: overloading operator < for comparing 2 Tags
        this is for priority queue's TagPtrComp comparision class
        want: smaller topLeftY, bigger H()
 * input: this and Tag & t
 * output: if this' priority < that of t's
 */
bool Tag :: operator < (Tag & t){
    return topLeftY() > t.topLeftY() ? true:
        topLeftY() < t.topLeftY() ? false:
        H() < t.H() ? true : false;/*H() > t.H() ? false:
        topLeftX() > t.topLeftX() ? true:
        topLeftX() < t.topLeftX() ? false:
        W() < t.W() ? true : false;*/
}

/*
 * name: expandRow
 * description: if t protrudes downwards, this tag will be wrapped with one 
        expanding to t's botRightY so this can include more tags as a row)
 * input: a pointer to another Tag
 * output: a pointer to the wrapper if wrapped, this otherwise
 */
Tag * Tag :: expandRow(Tag * t){
    //t is too long in y axis
    if(t -> botRightY() > botRightY()){ 
        return new Tag
            (topLeftY(), topLeftX(), t -> botRightY(), botRightX(), this, 0);
    }
    return this;
}

/* 
 * name: wrap
 * description: static, used to wrap 2 tags in a bigger parent tag, 
        and their positions are relative to their parent now.
 * input: 2 Tag * to be wrapped together
 * output: the parent wrapper
 */
Tag * Tag :: wrap(Tag * a, Tag * b){
    //smallest tly, tlx, largest bry, brx
    double tly = min(a -> topLeftY(), b -> topLeftY());
    double tlx = min(a -> topLeftX(), b -> topLeftX());
    double bry = max(a -> botRightY(), b -> botRightY());
    double brx = max(a -> botRightX(), b -> botRightX());
    Tag * t = new Tag(tly, tlx, bry, brx, a, b);
    //below are dirty css tricks
    //horizontal adjustment
    double ax = 100 * (a -> topLeftX() - t -> topLeftX()) / t -> W();
    if(ax > 0){
        a -> style += "\tleft: " + to_string(ax) + "%;\n";
    }
    double bx = 100 * (b -> topLeftX() - t -> topLeftX()) / t -> W();
    if(bx > 0){
        b -> style += "\tleft: " + to_string(bx) + "%;\n";
    }
    //vertical adjustment
    double ay = 100 * (a -> topLeftY() - t -> topLeftY()) / t -> H();
    if(ay > 0){
        a -> style += "\ttop: " + to_string(ay) + "%;\n";
    }
    double by = 100 * (b -> topLeftY() - t -> topLeftY() - a -> H()) / t -> H();
    if(by != 0){
        b -> style += "\ttop: " + to_string(by) + "%;\n";
    }
    return t;
}

/********************getters make life easier********************/

//returns opening tag to be written in html file
string Tag :: openTag(){
    string t = "<" + name;
    if(id != ""){
        t += " id = \"" + id + "\"";
    }
    if(cls != ""){
        t += " class = \"" + cls + "\"";
    }
    if(url != ""){
        if(name == "img"){
            return " src = \"" + url + "\">\n";
        }else{
            t += " href = \"" + url + "\"";
        }
    }
    return t + ">\n";
}

//returning closing tag to be written in html file
string Tag :: closeTag(){
    if(name == "img"){
        return "";
    }
    return "</" + name + ">\n";
}

//get top left corner's y coordinate
const double Tag :: topLeftY(){
    return tl.first;
}

//get top left corner's x coordinate
const double Tag :: topLeftX(){
    return tl.second;
}

//get bottom right corner's y coordinate
const double Tag :: botRightY(){
    return br.first;
}

//get bottom right corner's x coordinate
const double Tag :: botRightX(){
    return br.second;
}

//get size
double Tag :: S(){
    return (botRightX() - topLeftX()) * (botRightY() - topLeftY());
}

//get width
double Tag :: W(){
    return botRightX() - topLeftX();
}

//get height
double Tag :: H(){
    return botRightY() - topLeftY();
}

//get parent(useful for updating %)
Tag * Tag :: P(){
    return parent;
}

//eof

File: Tag.py

#Dongyao Zhu

'''
(0, 0)                   X
  |---------------------->
  | top left 
  |   ____________________
  |  |                    |
  |  |                    |
  |  |                    |
  |  |____________________|
  |                   bottom right
Y v
'''

class Tag:

    #name counter
	count = {'wrap' : 0}

	#no close
	voidTags = {'img', 'input'}

	def __init__(self, tlx = 0, tly = 0, brx = 0, bry = 0, name = 'div', \
		style = '', wrap = False, a = None, b = None):

		self.tlx = tlx
		self.tly = tly
		self.brx = brx
		self.bry = bry
		self.W = brx - tlx
		self.H = bry - tly
		self.S = self.W * self.H
		self.name = name
		self.style = "\tposition: relative;\n" + style
		self.children = []
		#creates an actual Tag
		if wrap == False:
			self.name = name
			self.cls = name
			if name in Tag.count:
				Tag.count[name] += 1
				self.id = name + str(Tag.count[name])
			else:
				Tag.count[name] = 1
				self.id = name + '1'
			self.wPct = self.W
			self.hPct = self.H
			if name == 'p' or name[0] == 'h':
				self.style += "\tmargin: 0%;\n\tpadding: 1px;\n"
				self.style += "\toverflow: scroll;\n\tcolor: white;\n"
				self.style += "\tborder: 1px white dashed;\n"
		#creates a wrapper Tag
		else:
			Tag.count['wrap'] += 1
			self.cls = 'wrap'
			self.id = 'wrap' + str(Tag.count['wrap'])
			self.children.append(a)
			a.wPct = 100 * a.W / self.W
			a.hPct = 100 * a.H / self.H
			self.wPct = self.W
			self.hPct = self.H
			if(b != None):
				self.children.append(b)
				b.wPct = 100 * b.W / self.W
				b.hPct = 100 * b.H / self.H
				
	def __lt__(self, t):
		#comparison for priority queue
		return True if self.tly < t.tly else False if self.tly > t.tly \
			else True if self.H > t.H else False

	def __str__(self):
		#for debug
		return self.name + ' - '  + self.id + ' - ' + self.cls + ': ' + \
			str(self.tlx // 1) + ', ' + str(self.tly // 1) + ', ' + \
			str(self.brx // 1) + ', ' + str(self.bry // 1)

	def expandRow(self, t):
		'''
		if t protrudes downwards, this tag will be wrapped with one 
	    expanding to t's botRightY so this can include more tags as a row)
	    '''
		if(t.bry > self.bry): 
			return Tag(self.tlx, self.tly, self.brx, t.bry, \
				'div', '', True, self, None)
		return self

	def wrap(a, b):
		'''
		used to wrap 2 tags in a bigger parent tag, 
		and their positions are relative to their parent now.
		'''
		#smallest tly, tlx, largest bry, brx
		t = Tag(min(a.tlx, b.tlx), min(a.tly, b.tly), \
			max(a.brx, b.brx), max(a.bry, b.bry), 'div', '', True, a, b)
		#below are dirty css tricks
		#horizontal adjustment
		ax = 100 * (a.tlx - t.tlx) / t.W
		if ax > 0:
			a.style += '\tleft: ' + str(ax) + '%;\n'
		bx = 100 * (b.tlx - t.tlx) / t.W
		if bx > 0:
			b.style += '\tleft: ' + str(bx) + '%;\n'
		#vertical adjustment
		ay = 100 * (a.tly - t.tly) / t.H
		if ay > 0:
			a.style += '\ttop: ' + str(ay) + '%;\n'
		by = 100 * (b.tly - t.tly - a.H) / t.H
		if by != 0:
			b.style += '\ttop: ' + str(by) + '%;\n'
		return t

	def openTag(self):
		#returns opening tag to be written in html file
		t = '<' + self.name
		if self.id != '':
			t += ' id = \"' + self.id + '\"'
		if self.cls != '':
			t += ' class = \"' + self.cls + '\"'
		if self.cls == 'button':
			t += ' onclick = "alert(\'Greetings!\')"click me'
		if self.cls == 'input':
			t += ' type = "text" placeholder = "is PHP the best language?"'
		return t + '>'

	def closeTag(self):
		#returns closing tag to be written in html file
		if self.name in Tag.voidTags:
			return '\n'
		return '</' + self.name + '>\n'

